include/pmacc/algorithms/GlobalReduce.hpp

```diff
@@ -35,43 +35,89 @@ namespace pmacc
-             * Must called from any mpi process. This function use global blocking mpi calls.
-             * Don't create a instance befor you have set you cuda device!
-             * @param isActive true if mpi rank should be part of reduce operation, else false
+             *
+             * @attention Must called from any mpi process. This function use global blocking mpi calls.
+             *
+             * @param isActive true if MPI rank should participate in the reduction, else false
              */
             void participate(bool isActive)
             {
+                isParticipating = isActive;
                 mpi_reduce.participate(isActive);
             }
 
+            /* defines if the result of the MPI operation is valid
+             *
+             * @tparam MPIMethod type of the reduction method
+             * @param method used reduction method e.g.,
+             *                reduceMethods::AllReduce, reduceMethods::Reduce
+             * @return if result of operator() is valid*/
+            template<class MPIMethod>
+            bool hasResult(const MPIMethod& method = ::pmacc::mpi::reduceMethods::AllReduce())
+            {
+                return mpi_reduce.hasResult(method);
+            }
+
             /* Reduce elements in global gpu memeory
              *
-             * @param func functor for reduce which takes two arguments, first argument is the source and get the new
-             * reduced value. Functor must specialize the function getMPI_Op.
-             * @param src a class or a pointer where the reduce algorithm can access the value by operator [] (one
-             * dimension access)
+             * @param func Binary functor for the reduction. First parameter is used as input and result value. Functor
+             * must specialize the function getMPI_Op.
+             * @param src A class or a pointer where the reduction algorithm can access the value by operator [] (one
+             * dimension access). The data must be located on the device.
              * @param n number of elements to reduce
              *
              * @return reduced value (same on every mpi instance)
+             *
+             * @{
              */
             template<class Functor, typename Src>
             typename traits::GetValueType<Src>::ValueType operator()(Functor func, Src src, uint32_t n)
+            {
+                return (*this)(func, src, n, ::pmacc::mpi::reduceMethods::AllReduce());
+            }
+
+            /**
+             * @tparam MPIMethod type of the reduction method
+             * @param method used reduction method e.g.,
+             *               reduceMethods::AllReduce, reduceMethods::Reduce
+             *               reduceMethods::AllReduce, reduceMethods::Reduce
+             */
+            template<class Functor, typename Src, class MPIMethod>
+            typename traits::GetValueType<Src>::ValueType operator()(
```

Comment:
Note: It was required to have two signatures of this method instead of

```C++
template<class Functor, typename Src, class MPIMethod>
typename traits::GetValueType<Src>::ValueType operator()(Functor func,
              Src src,
              uint32_t n,
              MPIMethod const& method = reduceMethods::AllReduce)
```

else the function was not callable without the method.

Reply:
could you please elaborate what you mean with this comment?

Reply:
In general, it would be possible to have only one implementation of the `operator()` but in this case, it is required to have two else I was not able to call the operator without the last parameter which should be replaced by the default in cases the user is not providing it. I pointed to this fact to avoid that the question is coming up during the review.